enhancements/installer/byo-custom-dns.md

```diff
@@ -0,0 +1,573 @@
+  - "@patrickdillon, for Installer aspects"
+  - "@Miciah, for Ingress Controller aspects"
+  - "@cybertron, for in-cluster DNS aspects"
+  - "@2uasimojo, for hive expertise"
+approvers:
+  - "@sdodson"
+  - "@patrickdillon"
+  - "@zaneb"
+  - "@Miciah"
+  - "@cybertron"
+
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@joelspeed"
+creation-date: 2023-04-27
+last-updated: 2023-04-27
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/CORS-1874
+replaces:
+  - "https://github.com/openshift/enhancements/pull/1276"
+---
+
+# BYO Custom DNS
+
+## Summary
+
+This feature provides OpenShift customers running on AWS an option to use a
+custom DNS solution that is different from the default cloud DNS. OpenShift does
+not have access to this custom DNS solution and hence is not expected to
+configure it.
+
+## Motivation
+
+Some customers do not want to use the DNS solution provided by the underlying
+cloud due to regulatory [ITAR](https://www.gov-relations.com/) or operational
+contraints. It is important for OpenShift to provide a way for these customers
+to use their own preferred DNS solution while supporting their IPI deployments.
+
+This external custom DNS solution would be responsible for DNS records for
+`api.<cluster domain>`, `api-int.<cluster domain>` and
+`*.apps.<cluster domain>`. The Cloud Load Balancers (LBs) are still the
+preferred way for providing LB services for `api`, `api-int` and `ingress` in
+public and private hosted zones.
+
+In addition, in OpenShift's Managed offering, the external DNS solution is also
+responsible for DNS records for required OpenShift resources including quay.io,
+registry.redhat.io, api.openshift.com, cloud management endpoints, etc. This
+enhancement proposal will not talk about how these entries are configured for
+the OpenShift Managed offering. 
+
+### User Stories
+
+"As a cloud admin, I want to run an OpenShift cluster on AWS with my custom DNS
+solution in place of the default cloud DNS."
+
+TODO: Is there anything else to add here?
+
+### Goals
+
+- Enable OpenShift customers to use their custom DNS solution instead of the
+default cloud solution.
+- Start with AWS and lay the groundwork to add this capability for other cloud
+platforms like Azure and GCP.
+- Continue using the default cloud LB service for API, API-Int and Ingress.
+
+### Non-Goals
+
+We will not use the [Kubernetes External DNS provider](https://github.com/openshift/external-dns)
+and/or the [OpenShift External DNS Operator](https://github.com/openshift/external-dns-operator)
+to program dns entries across varied DNS providers. The External DNS Operator
+relies on being able to program the customer's DNS solution. The customer DNS
+solutions in consideration here do not allow OpenShift to configure them.
```

Comment:
I don't think this solution needs to worry about the set of DNSs the External DNS operator can configure. If the DNS solution the customer wants to use is supported by External DNS operator and they are OK with OpenShift configuring it, they can still use it.
This solution is useful for customers that want to use a DNS solution that is currently not supported by the External DNS operator and/or they don't want OpenShift to configure their DNS. @makentenza can add details about the customers we are targeting with this solution.

Reply:
If we're able to use the External DNS Operator it could provide some substantial advantages. Consider if we moved API load balancers to be provisioned/managed by services created on bootstrap and ran the external dns operator on the bootstrap node. This would allow installer/cluster--provisioned LBs (customers don't need to BYO) and DNS records _for the subset of DNS providers supported by the external DNS operator_. We would need to determine some form of support for the DNS providers not included in the external DNS operator, but that could be either of the approaches we have already discussed (byo LBs or post-configured DNS).

 Abhinav did [a write up of SLB managed API load balancers](https://github.com/openshift/enhancements/pull/459), which might provide some more context about the approach or challenges. I wonder if using the external DNS operator would mitigate some (or most) of the issues he mentions.

Reply:
Looking at the feature in Jira, it says:

> While cloud-based DNS services provide convenient hostname management, there's a number of regulatory and operational constraints customers face prohibiting the use of those DNS hosting services on public cloud providers.

So, yeah, maybe that does knock out the option of using the external DNS operator.

Reply:
Please refer to https://issues.redhat.com/browse/OCPSTRAT-261 which states:

```
Out of Scope
The Installer to manage (create/update/delete) the required DNS records to deploy OpenShift on these cloud providers or integrate with those DNS services in any way
```

Reply:
Solving the use case where there is _no_ programmatic access to DNS is not optional. So we cannot _require_ the external DNS operator.

I guess that for users who are using the external DNS operator anyway, it would be nice if they didn't also have to manually create their own load balancer. But that's not an explicit goal of this work.

Potentially we could have an API design that would leave room for this if it ever actually comes up - e.g. if we had a DNSProvider field where the options are (some variant of) UserManaged vs. OpenShiftManaged then we could add an ExternalDNSOperatorManaged option if needed in the future. (Please don't use any of these names :slightly_smiling_face:)